From 0ce859577da4eccc86acdf94fbc4d7417da259ab Mon Sep 17 00:00:00 2001
From: Aleksander Morgado <aleksander@aleksander.es>
Date: Thu, 23 Feb 2017 10:53:38 +0100
Subject: [PATCH 2/2] build: musl compatibility

---
 src/libqmi-glib/qmi-utils.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/libqmi-glib/qmi-utils.c b/src/libqmi-glib/qmi-utils.c
index 5d399a6..8a059ba 100644
--- a/src/libqmi-glib/qmi-utils.c
+++ b/src/libqmi-glib/qmi-utils.c
@@ -740,7 +740,7 @@ __qmi_utils_get_driver (const gchar *cdc_wdm_path)
          *    /sys/bus/usb/drivers/qmi_wwan
          */
         tmp = g_strdup_printf ("/sys/class/%s/%s/device/driver", subsystems[i], device_basename);
-        path = canonicalize_file_name (tmp);
+        path = realpath (tmp, NULL);
         g_free (tmp);
 
         if (!path)
-- 
2.13.1

