--- a/include/net/dsa.h
+++ b/include/net/dsa.h
@@ -25,6 +25,7 @@
 	 */
 	struct device	*mii_bus;
 	int		sw_addr;
+	int		root;
 
 	/*
 	 * The names of the switch's ports.  Use "cpu" to
@@ -97,6 +98,7 @@
 	 * Parent switch tree, and switch index.
 	 */
 	struct dsa_switch_tree	*dst;
+	int			root;
 	int			index;
 
 	/*
--- a/net/dsa/dsa.c
+++ b/net/dsa/dsa.c
@@ -107,6 +107,7 @@
 		return ERR_PTR(-ENOMEM);
 
 	ds->dst = dst;
+	ds->root = pd->root;
 	ds->index = index;
 	ds->pd = dst->pd->chip + index;
 	ds->drv = drv;
--- a/net/dsa/slave.c
+++ b/net/dsa/slave.c
@@ -37,14 +37,17 @@
 
 void dsa_slave_mii_bus_init(struct dsa_switch *ds)
 {
+	struct mii_bus *master;
+
 	ds->slave_mii_bus->priv = (void *)ds;
 	ds->slave_mii_bus->name = "dsa slave smi";
 	ds->slave_mii_bus->read = dsa_slave_phy_read;
 	ds->slave_mii_bus->write = dsa_slave_phy_write;
-	snprintf(ds->slave_mii_bus->id, MII_BUS_ID_SIZE, "dsa-%d:%.2x",
-			ds->index, ds->pd->sw_addr);
-	ds->slave_mii_bus->mdio_level = ds->master_mii_bus->mdio_level + 1;
-	ds->slave_mii_bus->parent = &ds->master_mii_bus->dev;
+	master = ds->master_mii_bus;
+	ds->slave_mii_bus->mdio_level = master->mdio_level + 1;
+	ds->slave_mii_bus->parent = &master->dev;
+	snprintf(ds->slave_mii_bus->id, MII_BUS_ID_SIZE, "dsa-%d.%d:%.2x",
+		ds->root, ds->index, ds->pd->sw_addr);
 }
 
 
