--- a/mss_2013_10_31_33895/make/Makefile.linux.mb	2014-03-24 17:56:35.782972837 -0700
+++ b/mss_2013_10_31_33895/make/Makefile.linux.mb	2014-03-24 17:56:59.186968356 -0700
@@ -18,6 +18,7 @@
 #HW_ACCEL_DEV  = fsl8313
 RTOS          = __RTOS_LINUX__
 MB_OS         = linux
+MB_BUILDMODE  = optimized
 
 CC              = $(BCC_PREFIX)$(CROSS_COMPILE)gcc
 CXX             = $(BCC_PREFIX)$(CROSS_COMPILE)gcc
--- a/mss_2013_10_31_33895/make/Makefile.linux	2014-03-24 17:56:29.837285202 -0700
+++ b/mss_2013_10_31_33895/make/Makefile.linux	2014-04-24 13:43:13.917287627 -0700
@@ -26,7 +26,7 @@
 #LINKFLAGS  = -lrt
 LINKFLAGS  = -lrt -Wl,--no-as-needed -lpthread
 #CFLAGS    = -Os -D__RTOS_LINUX__ -Wall
-override CFLAGS += -D__RTOS_LINUX__ -g -Wall
+override CFLAGS += -D__RTOS_LINUX__ -g -O3 -Wall
 CXXFLAGS   = $(CFLAGS)
 LIBS       = $(LINKFLAGS)
 
--- a/mss_2013_10_31_33895/src/common/moptions.h	2014-03-31 17:20:47.029013756 -0700
+++ b/mss_2013_10_31_33895/src/common/moptions.h	2014-03-31 17:26:01.381006147 -0700
@@ -197,9 +197,6 @@
 #define __ENABLE_MOCANA_SUPPORT_FOR_NATIVE_STDLIB__
 #define __ENABLE_MOCANA_MEM_PART__
 #define __DISABLE_MOCANA_RAND_ENTROPY_THREADS__
-#define __ENABLE_ALL_DEBUGGING__
-#define __ENABLE_MOCANA_DEBUG_CONSOLE__
-#define __MOCANA_DUMP_CONSOLE_TO_STDOUT__
 #define __UAPI_DEF_IPPROTO_V6 0 /* Need for Kernel version > 3.12 */
 #endif /* __ENABLE_MOCANA_IPSEC_LINUX_EXAMPLE__ */
 
@@ -211,9 +208,6 @@
 #define __ENABLE_MOCANA_CERTIFICATE_SEARCH_SUPPORT__
 #define __ENABLE_MOCANA_EXAMPLES__
 #define __ENABLE_MOCANA_MEM_PART__
-#define __ENABLE_ALL_DEBUGGING__
-#define __ENABLE_MOCANA_DEBUG_CONSOLE__
-#define __MOCANA_DUMP_CONSOLE_TO_STDOUT__
 #endif /* __ENABLE_MOCANA_IKE_LINUX_EXAMPLE__ */
 
 #if defined( __ENABLE_MOCANA_DTLS_SERVER__ )
