Index: mocana-nanosec-5.8.0-33895/mss_2013_10_31_33895/src/ipsec/ipsec_protos.h
===================================================================
--- mocana-nanosec-5.8.0-33895/mss_2013_10_31_33895/src/ipsec/ipsec_protos.h	(revision 34320)
+++ mocana-nanosec-5.8.0-33895/mss_2013_10_31_33895/src/ipsec/ipsec_protos.h	(working copy)
@@ -316,6 +316,8 @@
 #ifndef IPPROTO_IPV6
 #define IPPROTO_IPV6        41  /* IPv6-in-IP tunnelling */
 #endif
+
+#if !(defined __KERNEL__ && (__UAPI_DEF_IPPROTO_V6 == 0) && !(defined __ENABLE_MOCANA_64_BIT__)) 
 #ifndef IPPROTO_ICMPV6
 #define IPPROTO_ICMPV6      58  /* ICMPv6 */
 #endif
@@ -335,8 +337,8 @@
 #ifndef IPPROTO_DSTOPTS
 #define IPPROTO_DSTOPTS     60  /* IPv6 destination options */
 #endif
+#endif
 
-
 #ifdef __cplusplus
 }
 #endif
Index: mocana-nanosec-5.8.0-33895/mss_2013_10_31_33895/src/common/moptions.h
===================================================================
--- mocana-nanosec-5.8.0-33895/mss_2013_10_31_33895/src/common/moptions.h	(revision 34320)
+++ mocana-nanosec-5.8.0-33895/mss_2013_10_31_33895/src/common/moptions.h	(working copy)
@@ -20,6 +20,14 @@
 
 /* @CODEGEN-START */
 /* @CODEGEN-END   */
+#if __GNUC__
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#define MOC_LITTLE_ENDIAN
+#endif
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+#define MOC_BIG_ENDIAN
+#endif
+#endif
 
 #if defined( __RTOS_DUMMY__ )
 #define __DUMMY_RTOS__
@@ -192,6 +200,7 @@
 #define __ENABLE_ALL_DEBUGGING__
 #define __ENABLE_MOCANA_DEBUG_CONSOLE__
 #define __MOCANA_DUMP_CONSOLE_TO_STDOUT__
+#define __UAPI_DEF_IPPROTO_V6 0 /* Need for Kernel version > 3.12 */
 #endif /* __ENABLE_MOCANA_IPSEC_LINUX_EXAMPLE__ */
 
 #ifdef __ENABLE_MOCANA_IKE_LINUX_EXAMPLE__
