diff -Naur a/drivers/net/ixgbe/base/ixgbe_x550.c b/drivers/net/ixgbe/base/ixgbe_x550.c
--- a/drivers/net/ixgbe/base/ixgbe_x550.c	2016-08-02 20:33:00.000000000 -0700
+++ b/drivers/net/ixgbe/base/ixgbe_x550.c	2016-09-22 16:14:51.991605976 -0700
@@ -1541,9 +1541,12 @@
	case ixgbe_sfp_type_1g_lx_core1:
		*linear = false;
		break;
-	case ixgbe_sfp_type_unknown:
+	// 1g copper SFP support;
	case ixgbe_sfp_type_1g_cu_core0:
	case ixgbe_sfp_type_1g_cu_core1:
+		*linear = false;
+		break;
+	case ixgbe_sfp_type_unknown:
	default:
		return IXGBE_ERR_SFP_NOT_SUPPORTED;
	}
@@ -1712,7 +1715,9 @@

		/* Check if 1G SFP module. */
		if (hw->phy.sfp_type == ixgbe_sfp_type_1g_sx_core0 ||
-		    hw->phy.sfp_type == ixgbe_sfp_type_1g_sx_core1
+		    hw->phy.sfp_type == ixgbe_sfp_type_1g_sx_core1
+		    || hw->phy.sfp_type == ixgbe_sfp_type_1g_cu_core0
+		    || hw->phy.sfp_type == ixgbe_sfp_type_1g_cu_core1
		    || hw->phy.sfp_type == ixgbe_sfp_type_1g_lx_core0 ||
		    hw->phy.sfp_type == ixgbe_sfp_type_1g_lx_core1) {
			*speed = IXGBE_LINK_SPEED_1GB_FULL;
