--- ubox-2013-11-19/kmodloader.c	2013-11-23 12:35:53.000000000 -0800
+++ openwrt-git/trunk/build_dir/target-i386_i486_eglibc-2.15/ubox-2013-11-19/kmodloader.c	2013-12-13 15:40:10.509859740 -0800
@@ -54,6 +54,7 @@
 	struct avl_node avl;
 
 	char *name;
+	char *opts;
 	char *depends;
 
 	int size;
@@ -170,6 +171,7 @@
 		return NULL;
 
 	m->avl.key = m->name = strcpy(_name, name);
+	m->opts = "";
 
 	if (depends) {
 		m->depends = strcpy(_dep, depends);
@@ -455,7 +457,7 @@
 		todo = 0;
 		avl_for_each_element(&modules, m, avl) {
 			if ((m->state == PROBE) && (!deps_available(m, 0))) {
-				if (!insert_module(get_module_path(m->name), "")) {
+				if (!insert_module(get_module_path(m->name), m->opts)) {
 					m->state = LOADED;
 					m->error = 0;
 					loaded++;
@@ -710,6 +712,8 @@
 				continue;
 
 			m->state = PROBE;
+			if(opts)
+				m->opts = strdup(opts);
 			if (basename(gl.gl_pathv[j])[0] - '0' <= 9)
 				load_modprobe();
 
